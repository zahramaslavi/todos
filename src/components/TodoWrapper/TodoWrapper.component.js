import React, { Component } from 'react';

import './TodoWrapper.component.scss';

import TodoFilter from '../TodoFilter/TodoFilter.component';
import AddTodo from '../AddTodo/AddTodo.component';
import TodoList from '../TodoList/TodoList.component';

export default class TodoWrapper extends Component {

    constructor(props) {
        super(props);

        const storedState = JSON.parse(localStorage.getItem('todosState'));

        // sample todo {description: 'test',completed: false, id: 1}
        if ( storedState ) {
            this.state = storedState 
        } else {
            this.state = {
                todos: [],
                hideCompleted: false,
                currentId: 0
            };
        }
    }

    componentDidUpdate() {
        localStorage.setItem('todosState', JSON.stringify(this.state));
    }

    handleAddTodo(todo) {
        const tempTodo = {...todo, id: this.state.currentId};
        const currentId = this.state.currentId + 1;
        this.setState({currentId: currentId});
        this.setState({todos: [...this.state.todos, tempTodo]});
    }

    handleDeleteTodo(id) {
        this.setState({
            todos: [...this.state.todos.filter(todo => todo.id !== id)]
        });
    }

    handleToggleTodo(id) {
        const tempTodos = [...this.state.todos];
        const todoIndex = this.state.todos.findIndex(todo => todo.id === id )
        tempTodos[todoIndex] = {...tempTodos[todoIndex], completed: !tempTodos[todoIndex].completed}
        this.setState({todos: tempTodos});
    }

    fileteredTodos() {
        return this.state.todos.filter(todo => {
                    if ( this.state.hideCompleted ) {
                        if ( !todo.completed ) {
                            return true;
                        } 

                        return false;
                    }

                    return true;
                });
    }

    render() { 
        return (

            <div className="todoWrapper">
                <TodoFilter 
                    todoCount={this.fileteredTodos().length}
                    hideCompleted={this.state.hideCompleted}
                    toggleHideCompleted={() => this.setState({hideCompleted: !this.state.hideCompleted})} 
                />

                <AddTodo
                    addTodo={(todo) => this.handleAddTodo(todo)}
                />

                <TodoList 
                    todos={this.fileteredTodos()}
                    handleDeleteTodo={(id) => this.handleDeleteTodo(id)} 
                    handleToggleComplete={(id) => this.handleToggleTodo(id)} 
                />
            </div>
        );
    }
}

// export default TodoWrapper;