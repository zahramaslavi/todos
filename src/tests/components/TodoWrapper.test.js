import React from 'react';
import { shallow } from 'enzyme';
import TodoWrapper from '../../components/TodoWrapper/TodoWrapper.component';

// const onSubmitSpy = jest.fn();
const wrapper = shallow(<TodoWrapper/>);

test('should test TodoWrapper component snapshot', () => {
    expect(wrapper).toMatchSnapshot();
});

//State change tests
wrapper.setState({
    todos: [
        {description: 'test',completed: false, id: 0},
        {description: 'test',completed: false, id: 1},
        {description: 'test',completed: false, id: 2},
    ],
    hideCompleted: false,
    currentId: 3
});

const todoFilter = wrapper.find('TodoFilter').first();
test('should test TodoWrapper component toggleHideCompleted of TodoFilter component', () => {
    const prevHideCompleted = wrapper.state().hideCompleted;

    todoFilter.props().toggleHideCompleted();
    expect(wrapper.state().hideCompleted).toBe(!prevHideCompleted);
});

const addTodo = wrapper.find('AddTodo').first();
test('should test Todos component addTodo of AddTodo', () => {
    const prevTodosLength = wrapper.state().todos.length;
    const prevCurrentId = wrapper.state().currentId;

    addTodo.props().addTodo();
    expect(wrapper.state().todos.length).toBe(prevTodosLength + 1);
    expect(wrapper.state().currentId).toBe(prevCurrentId + 1);
});

const todoList = wrapper.find('TodoList').first();
test('should test Todos component handleDeleteTodo of TodoList', () => {
    const prevTodosLength = wrapper.state().todos.length;

    todoList.props().handleDeleteTodo(0);
    expect(wrapper.state().todos.length).toBe(prevTodosLength - 1);
});

test('should test Todos component handleToggleComplete of TodoList', () => {
    const id = 1;
    const prevCompleted = wrapper.state().todos.find(todo => todo.id == id).completed;
    todoList.props().handleToggleComplete(id);

    expect(wrapper.state().todos.find(todo => todo.id == id).completed).toBe(!prevCompleted);
});