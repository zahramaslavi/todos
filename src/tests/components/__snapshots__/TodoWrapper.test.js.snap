// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test TodoWrapper component snapshot 1`] = `
<div
  className="todoWrapper"
>
  <TodoFilter
    hideCompleted={false}
    todoCount={3}
    toggleHideCompleted={[Function]}
  />
  <AddTodo
    addTodo={[Function]}
  />
  <TodoList
    handleDeleteTodo={[Function]}
    handleToggleComplete={[Function]}
    todos={
      Array [
        Object {
          "completed": false,
          "description": "test",
          "id": 0,
        },
        Object {
          "completed": false,
          "description": "test",
          "id": 1,
        },
        Object {
          "completed": false,
          "description": "test",
          "id": 2,
        },
      ]
    }
  />
</div>
`;
